File: zenscript-code-model/src/main/java/raylras/zen/model/type/IntType.java
package raylras.zen.model.type;

import raylras.zen.model.symbol.Operator;
import raylras.zen.model.symbol.Symbol;
import raylras.zen.model.symbol.SymbolFactory;

import java.util.List;

public enum IntType implements NumberType {

    INSTANCE;

    @Override
    public String getTypeName() {
        return "int";
    }

    @Override
    public List<Symbol> getSymbols() {
        return SymbolFactory.builtinSymbols()
                .add(NumberType.super.getSymbols())
                .operator(Operator.RANGE, IntRangeType.INSTANCE, params -> params.parameter("value", this))
                .build();
    }

    @Override
    public String toString() {
        return getTypeName();
    }

}

File: zenscript-code-model/src/main/kotlin/raylras/zen/model/type/IntType.kt
package raylras.zen.model.type

import raylras.zen.model.CompilationEnvironment
import raylras.zen.model.symbol.Operator
import raylras.zen.model.symbol.Symbol
import raylras.zen.model.symbol.symbolSequence

object IntType : NumberType {
    override val typeName = "int"

    override fun getSymbols(env: CompilationEnvironment?): Sequence<Symbol> {
        return symbolSequence {
            operator(Operator.RANGE, IntRangeType) {
                parameter("value", IntType)
            }
        } + super.getSymbols(env)
    }

    override fun toString() = typeName
}
